util/src/test/java/com/google/cloud/hadoop/util/TestRequestTracker.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2024 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.hadoop.util;
+
+import static com.google.common.truth.Truth.assertThat;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class TestRequestTracker extends RequestTracker {
+  private final List<GcsJsonApiEvent> events = new ArrayList<>();
+
+  public List<GcsJsonApiEvent> getEvents() {
+    return events;
+  }
+
+  @Override
+  protected void postToEventQueue(GcsJsonApiEvent event) {
+    this.events.add(event);
+  }
+
+  void verifyEvents(List<ExpectedEventDetails> expectedEvents) {
+    List<GcsJsonApiEvent> actualEvents = getEvents();
+    assertThat(actualEvents.size()).isEqualTo(expectedEvents.size());
+    assertThat(actualEvents.size()).isAtLeast(1);
+
+    for (int i = 0; i < actualEvents.size(); i++) {
+      GcsJsonApiEvent actual = actualEvents.get(i);
+      ExpectedEventDetails expected = expectedEvents.get(i);
+
+      assertThat(actual.getEventType()).isEqualTo(expected.eventType);
+      assertThat(actual.getContext().toString()).isEqualTo(expected.context.toString());
+
+      GcsJsonApiEvent.EventType eventType = actual.getEventType();
+      if (eventType == GcsJsonApiEvent.EventType.RESPONSE) {
+        verifyNotEmpty(actual, GcsJsonApiEvent.DURATION);
+      }
+
+      if (eventType == GcsJsonApiEvent.EventType.RESPONSE) {
+        verifyNotEmpty(actual, GcsJsonApiEvent.DURATION);
+      }
+
+      if (eventType == GcsJsonApiEvent.EventType.BACKOFF) {
+        verifyNotEmpty(actual, GcsJsonApiEvent.BACKOFFTIME);
+      }
+
+      for (String key : expected.properties.keySet()) {
+        if (key.equals(GcsJsonApiEvent.BACKOFFTIME)) {
+          long backOffTime = (long) actual.getProperty(key);
+          int expectedBackoffTime = (int) expected.properties.get(GcsJsonApiEvent.BACKOFFTIME);
+          assertThat(backOffTime).isAtLeast(expectedBackoffTime);
+          assertThat(backOffTime).isLessThan(expectedBackoffTime + 10);
```

Comment:
why it varies as range? 
How we came up with `10`?
will it cause flakiness? 
what if some adds a test where many more retires are done, this verification will still be valid?

Reply:
These are unit tests, not actual backoff is happening. The buffer is the potential overhead. I do not expect it to be more than 10 seconds.

Reply:
> These are unit tests, not actual backoff is happening.

Well, that is not always true, is it? One can mock the transport in unit tests but still use RetryHttpIniitalizer which comes with ExponentialBackOff strategy and hence the actual backoff will will happen.